### PR TITLE
implemented adaptive scaling

### PR DESCRIPTION
**Pull request by sinamna**
adaptively scale based on target deployment

## Summary by CodeRabbit

- **New Features**
  - Added a search functionality to the app.

- **Style**
  - Added styles for the search bar.

- **Refactor**
  - Introduced a new `Search` class with methods for fetching and displaying results.
  - Added a search bar to the `Hero` component and a `Search` component to the `App` component.

### CODE REVIEW

File: internal/controller/basic_authenticator/utility.go [hunk 1]
```diff
@@ -1,10 +1,25 @@
...
+func getServiceName(appService string) string {
+	serviceParts := strings.Split(appService, ".")
+	if len(serviceParts) == 3 && serviceParts[2] == "svc" {
+		// checking for appservice.ns.svc format
+		return serviceParts[0]
+	} else if len(serviceParts) == 1 {
+		return serviceParts[0]
+	} else {
+		return ""
+	}
```
[the remark this answers is not in the document]
Comment (sinamna) [reply internal/controller/basic_authenticator/utility.go:24]: its not needed. ill delete it

File: internal/controller/basic_authenticator/provision.go [hunk 2]
```diff
@@ -251,3 +290,39 @@ func (r *BasicAuthenticatorReconciler) CreateSidecarAuthenticator(ctx context.Co
...
+	err := r.Update(ctx, &targetDeploy)
+	if err != nil {
+		return -1, err
+	}
+	replicas := deployments.Items[0].Spec.Replicas
+	targetReplica := math.Floor(float64((*replicas + 1) / 2))
+	log.FromContext(context.Background()).Info("-------debug", "target Replica inside", targetReplica)
+
+	return int32(targetReplica), nil
+}
```
[the remark this answers is not in the document]
Comment (sinamna) [reply internal/controller/basic_authenticator/provision.go:328]: provide a explanation to include in crd definition for appService that it should select only one deployment in the same namespace as basic authenticator and should be only kubernetes service name

File: internal/controller/basic_authenticator/provision.go [hunk 3]
```diff
@@ -210,15 +226,38 @@ func (r *BasicAuthenticatorReconciler) CreateDeploymentAuthenticator(ctx context
...
 			}
+			err = r.Get(ctx, types.NamespacedName{Name: foundDeployment.Name, Namespace: basicAuthenticator.Namespace}, foundDeployment)
+			if err != nil {
+				logger.Error(err, "failed to refetch")
+				return subreconciler.RequeueWithError(err)
+			}
+
 		}
 		logger.Info("updating ready replicas")
 		basicAuthenticator.Status.ReadyReplicas = int(foundDeployment.Status.ReadyReplicas)
```
[the remark this answers is not in the document]
Comment (sinamna) [reply internal/controller/basic_authenticator/provision.go:263]: ill remove it in next commits

File: internal/controller/basic_authenticator/basicauthenticator_controller.go [hunk 4]
```diff
@@ -56,5 +61,29 @@ func (r *BasicAuthenticatorReconciler) SetupWithManager(mgr ctrl.Manager) error
...
+	if !exists {
+		return nil
+	}
+	log.FromContext(context.Background()).Info("deployment before revision", "revision", deploy.ResourceVersion, "basic Auth name", basicAuthName, "ns", deploy.Namespace)
+	return []reconcile.Request{
+		{
+			NamespacedName: types.NamespacedName{Name: basicAuthName, Namespace: deploy.Namespace},
+		},
+	}
+}
```
[the remark this answers is not in the document]
Comment (sinamna) [reply internal/controller/basic_authenticator/basicauthenticator_controller.go:89]: the comment is not necessary. it'll be removed

**Comment by hoptical** [comment]
`CreateCredentials` in `nginx.go` is not needed to be a method of `BasicAuthenticatorReconciler`. 
Additionally, all functions in `nginx.go` can be private.

**Comment by hoptical** [comment]
The [created secret](https://github.com/snapp-incubator/simple-authenticator/blob/024592012714a38c6fc69ca6fae8c1a56e8ecd16/internal/controller/basic_authenticator/provision.go#L66-L67) needs a controller reference.

**Comment by hoptical** [comment]
[These lines](https://github.com/snapp-incubator/simple-authenticator/blob/5d406b0fa3b461bd74dc9502ce934db2b4bfdbfb/internal/controller/basic_authenticator/nginx.go#L159-L167) in injector function are common with createNginxDeployment function. Can be moved out to another function.

**Comment by sinamna** [comment]
> The [created secret](https://github.com/snapp-incubator/simple-authenticator/blob/024592012714a38c6fc69ca6fae8c1a56e8ecd16/internal/controller/basic_authenticator/provision.go#L66-L67) needs a controller reference.

done

File: pkg/random_generator/random_generator.go [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+)
+
+func GenerateRandomString(length int) string {
+	const charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+
+	randomBytes := make([]byte, length)
+	rand.Read(randomBytes)
```
[the remark this answers is not in the document]
Comment (sinamna) [reply pkg/random_generator/random_generator.go]: done

File: internal/controller/basic_authenticator/provision.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (hoptical) [reply internal/controller/basic_authenticator/provision.go]: @sinamna What's your decision about this?

File: internal/controller/basic_authenticator/provision.go
[the remark this answers is not in the document]
Comment (hoptical) [reply internal/controller/basic_authenticator/provision.go:1]: @sinamna

[the remark this answers is not in the document]
Comment (hoptical) [reply internal/controller/basic_authenticator/provision.go:1]: @sinamna

[the remark this answers is not in the document]
Comment (sinamna) [reply internal/controller/basic_authenticator/provision.go:1]: based on [redhat introduced pattern](https://cloud.redhat.com/blog/subreconciler-patterns-in-action#:~:text=*MemcachedReconciler)-,getLatestMemached(,-ctx%20context.) intended to call getLatest when each subreconciler was called. not on each reconcile loop.

[the remark this answers is not in the document]
Comment (hoptical) [reply internal/controller/basic_authenticator/provision.go:1]: Agree

[the remark this answers is not in the document]
Comment (sinamna) [reply internal/controller/basic_authenticator/provision.go:1]: done

**Review by hoptical — APPROVED** [review]

**Pull request merged**